crypto/zk/utils.go

```diff
@@ -2,61 +2,119 @@
 package zk
```

Comment:
my brain dumped at this point, will finish reviewing next week, starting from here

Reply:
Also here.